### ISSUE TITLE
download-queue-size or seed-queue-size issue

### ISSUE THREAD

**Issue by ejunjsh** [transmission/transmission#5938]
### What is the issue?

i have 5 downloads , and download-queue-size is 5, but when i set download-queue-size to 4, the 5 downloads still in downloading status, is it ok? or one of them should be change to queued?

i'm sure they are active, not stall.

but when i restart them, 4 is running , 1 is queued, so when i reduce download-queue-size , i need to restart the downloads?

the issue happen on seed-queue-size too

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.3

**Comment by killemov** [comment]
This is by design. In your example of q-size 4, when one of the 4 downloading torrents becomes stalled the 5th torrent will start. In [Shift](https://github.com/killemov/Shift) there is an option "Start Now" to skip the q.

**Comment by tearfur** [comment]
@ejunjsh Do you mean you changed `settings.json` while Transmission was running? Changes to `settings.json` will only take effect if Transmission was not running when you changed it.

**Comment by ejunjsh** [comment]
> @ejunjsh Do you mean you changed `settings.json` while Transmission was running? Changes to `settings.json` will only take effect if Transmission was not running when you changed it.

thanks for reply, i changed settings by rpc, so i think  the change will take effect immediately

i check the code  `queuePulse` this method to do the check when queue change,  but the method only handle the case when queue have free slot, but when queue size is less than the running (active) downloads, it doesn't  put some downloads to queue.  is this by design?

**Comment by tearfur** [comment]
Looks more like an oversight to me more than by design. Thanks for reporting.

**Comment by killemov** [comment]
The thing with the settings.json got conflated with the actual queue thing (problem?). Is it an oversight to keep running torrents running when you decrease the queue-size? (That is what I'm getting from @tearfur .) AFAIK the stalled torrents will get to waiting state after a while when there are more torrents active than the queue-size. (I guess the naming could be done better as the number pertains to active torrents after which torrents are queued, correct?)

**Comment by tearfur** [comment]
> The thing with the settings.json got conflated with the actual queue thing

No, the issue author made it clear that this is about the RPC interface.

> Is it an oversight to keep running torrents running when you decrease the queue-size?

I think so. The `settings.json` docs made it very clear: "When true, Transmission will only download `download-queue-size` non-stalled torrents at once."

So when there is an interface provided for you to change the parameter live (a.k.a. the RPC interface), I'd expect it to take effect immediately.

**Comment by tearfur** [comment]
Removed the `good first issue` label after I did some preliminary assessment. You'd have to mess with a lot of Transmission's most fundamental groundwork to get this working, and it is not easy to not get it wrong.

**Comment by killemov** [comment]
@tearfur wrote:
> I think so. The `settings.json` docs made it very clear: "When true, Transmission will only download `download-queue-size` non-stalled torrents at once."

Ah, if you wish to change the code to fit the documentation then yes. In this case I would strongly advise to change the documentation to fit the code. Would the process be to simply queue the torrent with the highest queue-id? What if that is the one currently downloading the most?

**Comment by tearfur** [comment]
> Would the process be to simply queue the torrent with the highest queue-id?

Yes, this is what I would do to fit into the current design of the queuing feature.

> What if that is the one currently downloading the most?

I don't use the queues in the first place so I don't really know what's best. Always keeping only the most "high-performing" torrents might starve the others, and I am not convinced this is what you'd want. That's why there is a feature to modify the queue order, so people can choose the torrents most important to them (although it is broken right now).

In any case, looking at the code, it seems like transfer performance was not factored into the default queue order in the first place. If you want the feature to factor that in, it'd have to be overhauled.

You're welcome to open PRs, start discussions about it and whatnot if this is a high priority to you.
